[stdlib] Rewriting native hashed collection indices (rebased)

Gankra opened the pull request:
<!-- What's in this pull request? -->

This is a rebase of #3046. The majority of the work here is @austinzheng's, and can be found in the first commit. However the codebase had changed enough that the patch had to be significantly modified. I'm mostly relying on the test suite to verify that there weren't errors in resolving all the conflicts.

```
Changes:
- Native dictionary and set indices no longer hold references to storage
- Cocoa-based dictionary and set indices no longer hold references to storage
- Removed double indirection trick from hashed collections
- Rewrote storage types to reflect simpler model
- Updated unit tests
```

As a consequence of removing the double indirection, Dictionary -> NSDictionary now always requires an allocation for a wrapper class. This means that if you bridge a Dictionary multiple times, the resultant NSDictionaries will not be referentially identical. I briefly investigated trying to restore the previous behaviour; or at least perform the optimizations that Array does to make the verbatim-bridgeable case free. 

Discussion with @dabrahams lead to the conclusion that restoring the previous behaviour isn't necessary or even strictly desirable. In particular, the old behaviour could lead to the bridged buffer being kept alive much longer than necessary. This design ensures those are cleaned up as soon as possible. 

My own research into adopting Array's optimization appears to be infeasible, as _NativeDictionaryStorage already inherits from ManagedBuffer, and so can't inherit from _NSDictionaryCore as well (right?).

Still, it should be noted that this will almost certainly lead to regressions in bridging performance. At least in micro-benchmarks. 

It should also be noted that this redesign is necessary for soundness, as the double-indirection design is unsound for multi-threading (as discussed in SE-0065: https://github.com/apple/swift-evolution/blob/master/proposals/0065-collections-move-indices.md). 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: 
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN). -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

TODO before merge:
- [x] re-instate empty-collection-singleton
- [x] fix comparison change
- [ ] evaluate benches
- [x] fix documentation

Commit:
[stdlib] Rewriting native hashed collection indices

Changes:
- Native dictionary and set indices no longer hold references to storage
- Cocoa-based dictionary and set indices no longer hold references to storage
- Removed double indirection trick from hashed collections
- Rewrote storage types to reflect simpler model
- Updated unit tests

Commit:
[stdlib] Update Dictionary comments to reflect new design

Commit:
[stdlib] Handle some outstanding Dictionary indexing-model FIXMEs/TODOs

Gankro commented:
(repearing this comment from the other PR)

The only outstanding problem is that the new bridging implementation seems to be breaking equality comparisons for [AnyHashable: Hashable] when bridged to NSDictionary. In particular, this assert is now failing: https://github.com/apple/swift/blob/c3b7709a7c4789f1ad7249d357f69509fb8be731/test/stdlib/TestUserInfo.swift#L101

I'm guessing this is because this kind of Dictionary doesn't provide its own equality comparison, that it's falling back to some object identity stuff, which is no longer holding? Not sure what guarantees we're providing here, or how to proceed. Any idea what the right solution is?

CC @moiseev @dabrahams

Gankro commented:
@swift-ci please smoke test

Gankro commented:
@swift-ci please benchmark

Gankro commented:
Oh interesting. The new source-stability machinery caught that we removed Index.successor(). I genuinely have no idea how we can source-stability bridge that.

Commit:
update reflection tests with new type layouts for Dictionary

dabrahams reviewed (commented):
Is this equality "dubious," or simply no longer something we can keep providing once bridging is made eager?

Gankro commented:
@dabrahams to be honest, I don't even understand how it was passing before. Why would two independently constructed empty dictionaries have the same address?

Gankro commented:
Wait are you referring to the comment I made above, or the last commit I pushed?

Gankro commented:
These bench results seem pretty promising! Major regressions are in the expected place (bridging), and there's some big wins on a few native benchmarks!

dabrahams commented:
@bob-wilson Can someone on your team evaluate the performance results?

Gankro commented:
@jrose-apple I'm told you're the right person to ask about the correct behaviour for https://github.com/apple/swift/pull/5291#issuecomment-253812855

jrose-apple commented:
I can't see how it ever could have been depending on object identity, but those two dictionaries definitely should remain equal. They're being created from the same source.

(It's possible, of course, that something is broken in either AnyHashable or our SwiftValue box, but then I'd expect it to be broken everywhere.)

@jckarter is the other person to ask here (especially without Dmitri). Any insights, Joe?

jrose-apple commented:
Ah, it could conceivably be fallout from @DougGregor's https://github.com/apple/swift/commit/25e00d78d14cc174bfe924ed28c3355a33e22e59 as well.

Gankro commented:
The source-stability checker seems to be a false positive. `.successor()` is internal and unversioned. There shouldn't be any problem in removing it. How do we assert this to the system? (Just ignore it and manually merge when everything else is resolved?)

Gankro commented:
Reverting https://github.com/apple/swift/commit/25e00d78d14cc174bfe924ed28c3355a33e22e59 doesn't seem to affect this test. 

@jrose-apple how does the `==` implementation for NSDictionary even come into existence? What features of our implementation of NSDictionary are used to synthesize it? Genuinely can't find this information.

jrose-apple commented:
`==` for NSDictionary is provided by `==` for NSObject, which invokes `isEqual(_:)`, which calls NSDictionary's override of that, which should check "contains same key-value pairs", which will call `isEqual(_:)` on keys and values. In this case the keys and values aren't Objective-C types, so I would expect them to be boxed, but it's also stretching beyond any areas I worked on. (Hence poking Joe.)

Gankro commented:
Ok yeah, that makes sense. I can see why this was working before and now it's broken.

Here's a minimal example demonstrating the problem:

```swift
import Foundation

struct MyType {
    var data: Int = 0
}

let dictOfEqs: [AnyHashable: Any] = [AnyHashable(1): 1]
let dictOfMisc: [AnyHashable: Any] = [AnyHashable(1): MyType()]
let dictOfDupe: [AnyHashable: Any] = [AnyHashable(1): MyType()]


let nsDictOfEqs1 = dictOfEqs as NSDictionary
let nsDictOfEqs2 = dictOfEqs as NSDictionary

print(nsDictOfEqs1 == nsDictOfEqs2)


let nsDictOfMisc1 = dictOfMisc as NSDictionary
let nsDictOfMisc2 = dictOfMisc as NSDictionary

print(nsDictOfMisc1 == nsDictOfMisc2)

let nsDictOfDupe = dictOfDupe as NSDictionary

print(nsDictOfMisc1 == nsDictOfDupe)
```

In Swift 3, this code prints:

```
true
true
false
```

With this patch, we get:

```
true
false
false
```

Why? Before, Dictionary cached the bridged buffer. So multiple bridgings stored the exact same elements if they had to be boxed. This is relevant for types which aren't natively comparable in Swift, because ObjC is forced to fall back to pointer identity. With a cached buffer, pointer identity holds for a dictionary that has been bridged multiple times!

*But* this is easily broken by making two Dictionaries which don't share the same cache, so this behaviour isn't especially reliable (hence why we always get false for the last case).

I would argue that while the Swift 3 behaviour is *nice*, it's arguably incorrect. Or at best, a foot gun. A model where `as` is understood as "pass by value", and therefore makes values with distinct identity is an easier to understand one.

Note that with this patch `[AnyHashable: Any]` is still well behaved if the Any'd type provides its own implementation of `==`.  This only breaks people relying on two non-comparable values having equal boxings.

jrose-apple commented:
Ah, that makes sense!

It's (unfortunately) a requirement of NSDictionary and NSArray that getting elements repeatedly results in the same pointer value, which is why that cache was there at all. But there's no reason to assume the cache would be shared between elements, and it's certainly a coincidence to be able to do it at all. The same would apply in Objective-C to value-like types that don't override `-isEqual:`, so this seems like reasonable behavior.

Just to check, if you make MyType Equatable, we get the first result, right?

Gankro commented:
Huh, seemingly no? I had assumed that Any was peeking into that. Need to investigate why ints are doing magic to Any...

dabrahams commented:
> I would argue that while the Swift 3 behaviour is *nice*, it's
> arguably incorrect. Or at best, a foot gun. 

No, this is exactly what I explained to you was happening with
_SwiftDeferredNSArray, just repeated for dictionaries.  It would be
“nice” if we could get away with ignoring Cocoa programmers'
expectations that they get back the same objects (by identity) every
time they traverse an immutable collection, but, at least today, we
can't.

-- 
-Dave

jckarter commented:
I agree with @Gankro. I think this behavior change is unavoidable with eager bridging. Since `MyType` isn't `Equatable`, there's no equality behavior we can offer in the bridged `NSDictionary` beyond object identity, and there's no way to preserve object identity if we build a fresh `NSDictionary` every time we bridge. If we cached the `NSDictionary` in the `Dictionary` value somehow, that's just pushing the problem around, since the `Dictionary` in Swift is a value type and we can't guarantee any instance-specific behavior.

Gankro commented:
@dabrahams _SwiftDeferredNSArray has the behaviour this patch is introducing.

```
import Foundation

struct MyType {
    var data: Int = 0

    static func ==(lhs: MyType, rhs: MyType) -> Bool {
        return lhs.data == rhs.data
    }
}

let arrayOfEqs: [Any] = [1]
let arrayOfMisc: [Any] = [MyType()]
let arrayOfDupe: [Any] = [MyType()]

let nsArrayOfEqs1 = arrayOfEqs as NSArray
let nsArrayOfEqs2 = arrayOfEqs as NSArray

print(nsArrayOfEqs1 == nsArrayOfEqs2)

let nsArrayOfMisc1 = arrayOfMisc as NSArray
let nsArrayOfMisc2 = arrayOfMisc as NSArray

print(nsArrayOfMisc1 == nsArrayOfMisc2)

let nsArrayOfDupe = arrayOfDupe as NSArray

print(nsArrayOfMisc1 == nsArrayOfDupe)
```

yields

```
true
false
false
```

We ensure that once bridged, an NSDictionary yields stable addresses. What NSDictionary is currently doing is ensuring that *different NSDictionarys bridged from the same Dictionary* have stable addresses.

Gankro commented:
Can anyone direct me to where the isEqual impl for Any/AnyObject is constructed? I can't figure out why Int's `==` impl is being passed through, but MyType's isn't.

jckarter commented:
@Gankro Look at `SwiftValue.mm`. AFAICT, in your example `MyType` doesn't *have* an Equatable conformance, so the `_SwiftValue` bridging box ought to fall back to object identity. This is all behaving correctly AFAICT.

Gankro commented:
@jckarter yeah my posted examples don't, but adding the conformance doesn't seem to affect the behaviour of the code:

```swift
import Foundation

struct MyType: Equatable {
    var data: Int = 0

    static func ==(lhs: MyType, rhs: MyType) -> Bool {
        return lhs.data == rhs.data
    }
}



let dictOfEqs: [AnyHashable: Any] = [AnyHashable(1): 1]
let dictOfEqsDupe: [AnyHashable: Any] = [AnyHashable(1): 2]
let dictOfMisc: [AnyHashable: Any] = [AnyHashable(1): MyType()]
let dictOfDupe: [AnyHashable: Any] = [AnyHashable(1): MyType()]


let nsDictOfEqs1 = dictOfEqs as NSDictionary
let nsDictOfEqs2 = dictOfEqs as NSDictionary

print(nsDictOfEqs1 == nsDictOfEqs2)

let nsDictOfEqsDupe = dictOfEqsDupe as NSDictionary

print(nsDictOfEqs1 == nsDictOfEqsDupe)

let nsDictOfMisc1 = dictOfMisc as NSDictionary
let nsDictOfMisc2 = dictOfMisc as NSDictionary

print(nsDictOfMisc1 == nsDictOfMisc2)

let nsDictOfDupe = dictOfDupe as NSDictionary

print(nsDictOfMisc1 == nsDictOfDupe)

print("\narrays:")


let arrayOfEqs: [Any] = [1]
let arrayOfMisc: [Any] = [MyType()]
let arrayOfDupe: [Any] = [MyType()]

let nsArrayOfEqs1 = arrayOfEqs as NSArray
let nsArrayOfEqs2 = arrayOfEqs as NSArray

print(nsArrayOfEqs1 == nsArrayOfEqs2)

let nsArrayOfMisc1 = arrayOfMisc as NSArray
let nsArrayOfMisc2 = arrayOfMisc as NSArray

print(nsArrayOfMisc1 == nsArrayOfMisc2)

let nsArrayOfDupe = arrayOfDupe as NSArray

print(nsArrayOfMisc1 == nsArrayOfDupe)
```

before patch:

```
true
false
true
false

arrays:
true
false
false
```

after patch:

```
true
false
false
false

arrays:
true
false
false
```

Gankro commented:
@_@ the isEqual impl uses Hashable, not Equatable

dabrahams commented:
on Mon Oct 17 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> We ensure that once bridged, an NSDictionary yields stable
> addresses. What NSDictionary is currently doing is ensuring that
> *different NSDictionarys bridged from the same Dictionary* have stable
> addresses.

Oh!  In that case, yeah, kill that behavior.  Nobody asked for that
behavior AFAIK.

-- 
-Dave

dabrahams commented:
on Mon Oct 17 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

>  @_@ the isEqual impl uses Hashable, not Equatable

Arguably we ought to have an AnyEquatable, too, but that's a separate
project.

-- 
-Dave

Gankro commented:
Looks like rdar://problem/27574348 (Forward -isEqual: and -hash through SwiftObject, SwiftValue, and SwiftNativeNSError) was never fully implemented. I can't find any discussion of why only the Hashable conformance was used rather than Equatable here.

But it seems like we can just ignore that as an unrelated bug for now.

Commit:
properly expose == to objc in test

This test was relying on an impl detail that was never guaranteed and
no longer holds (multiple bridgings of a dict had equal identity). Based
on how the code in the test is written, it appears to be an accident.
This change makes the test "make sense" without the impl detail.

Commit:
refactor Dictionary to support an empty singleton

Gankro commented:
@dabrahams I've added the logic to include an empty singleton, but after doing that I realized the failing test is actually verifying an empty set *bridged* NSDictionary is also a singleton. Unfortunately, as discussed this no longer comes free with Dictionary, so I'd need to further refactor _NativeSelfBridgingStorage to also be a class hierarchy like I have with _NativeSelfStorage.

Before I proceed I'd like to verify that:

* My solution to this problem is acceptable
* We care about bridging empty sets/dictionaries enough to bother

dabrahams commented:
on Mon Oct 17 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> @dabrahams I've added the logic to include an empty singleton, but
> after doing that I realized the failing test is actually verifying an
> empty set *bridged* NSDictionary is also a singleton. Unfortunately,
> as discussed this no longer comes free with Dictionary, so I'd need to
> further refactor _NativeSelfBridgingStorage to also be a class
> hierarchy like I have with _NativeSelfStorage.
>
> Before I proceed I'd like to verify that:
>
> * My solution to this problem is acceptable

I'm afraid I don't know what the problem or your solution is.  Is there
any reason you can think of that it might be unacceptable?  If this is
more complicated, we should talk; I should be in the office around 10:30
PST.

> * We care about bridging empty sets/dictionaries enough to bother

I think the empty case is common enough to be worth optimizing for.
Normally what we'd do is make our singleton buffer also be-a
NSDictionary.

-- 
-Dave

Gankro commented:
I'm mostly just unhappy with this solution because a bunch of type-system crud is in the way and making it more complex/dynamic than it should be. I don't have any specific concerns with the design, it seems like the best one given the constraints. But I'm not a swift expert so I just want to verify that it actually is.

Current design:

* NativeRawBuffer (class) provides an untyped header/buffer
* NativeBuffer (class) inherits this to add enough type info to implement deinit
* NativeStorage stores a NativeRawBuffer, also providing type info, and all actual functionality
* 99% of the time, NativeStorage stores a NativeBuffer. However, the empty singleton is a "pure" NativeRawBuffer. Using a NativeRawBuffer is necessary to satisfy strict aliasing, because we want to reinterpret its absence of elements as whatever type the Dictionary has.

In my ideal world, either:

* NativeBuffer doesn't exist. It encodes redundant type info, and bloats up the runtime's class metadata system. Whenever a Dictionary is destroyed, it just checks for uniqueness of its buffer, and if true, runs the deinit it provides. This is currently impossible, as structs can't provide deinits. I doubt this will ever be possible, because even if we opt to provide them on move-only types, I doubt we'll provide them on copyable structs.

*  NativeRawBuffer doesn't exist. It exists only to satisfy strict aliasing. Type safety of the stored elements would be well-preserved by having the singleton be a `NativeBuffer<(), ()>` that we cast to `NativeBuffer<K, V>`, because there's no way to actually read/write keys/values to the empty singleton (unclear: whether this would melt some metadata system regardless of strict aliasing). I doubt this will ever be possible, because strict aliasing is pretty clearly here to stay.

atrick commented:
FWIW, I think NativeRawBuffer is the right solution to the singleton-which-may-standin-for-any-type problem. It might be better spelled NativeBufferBase or something...

jckarter commented:
It would be nice if we could represent a non-protocol existential like `exists T, U. NativeBuffer<T, U>` in the type system, which would avoid the need for the formal base class.

Commit:
refactor HashedCollections to support verbatim bridging

Gankro commented:
@swift-ci Please benchmark

Gankro commented:
@swift-ci Please smoke test

Gankro commented:
OK!

So here's the new design. After digging through how Array actually accomplishes what it does, I realized I had a benign but deceptive rebase error. (I was inheriting from ManagedBuffer *and* manually tail-allocating self). Removing that error resolved the double-inheritance issue, allowing verbatim bridging to be a thing!

Changes to design:

There's a new typed inserted between RawNativeBuffer and `NativeBuffer<K: Hashable, V>` -- `TypedBuffer<K, V>`. The purpose of this type is literally to provide typing without the Hashable bound. This is necessary to support `NativeStorage<AnyObject, AnyObject>`, which is now used by `SwiftDeferredNSDictionary<K, V>` (previously called `Native${Self}BridgingStorage`, which I thought was confusing).

RawNativeBuffer inherits from NSDictionary, but only providing code for "nothing is contained". NativeBuffer replaces this implementation with a proper one. Note that TypedNativeBuffer can't do this, because you can't mess with objc methods in constrained conformances (and the "real" implementation needs K: Hashable).

If K and V are verbatim bridgeable, or the backing store is the empty singleton, bridging to ObjC just returns the buffer. Toll free bridging!

Otherwise, we toss the `NativeStorage<K, V>` into SwiftDeferredNSDictionary, which will construct a bridged `NativeStorage<AnyObject, AnyObject>` on first use.

Some miscellaneous cleanup: 

* `BridgedNativeDictionaryStorage`, which was just a clone of NativeStorage with `<K, V>` hardcoded to `<AnyObject, AnyObject>` has been completely deleted. NativeStorage is the only remaining "storage" type.

* `NativeDictionaryStorageKeyNSEnumerator` has been renamed to `NativeDictionaryNSEnumerator`, because I hated that name (very Swift 2.0-ish).

* SwiftDeferredNSDictionary has been significantly simplified, because there is no longer a "verbatim bridgeable" fastpath; nothing verbatim bridgeable should ever be fed to it.

* Some conveniences for getting (bridged) keys/values were added to NativeStorage to deduplicate some logic.

Gankro commented:
Locally I'm seeing this assertion failing:

https://github.com/apple/swift/blob/master/validation-test/stdlib/Set.swift#L2047

Digging into it, I don't really understand what guarantee it's trying to uphold. To be clear, it appears to be testing that two empty Actually-From-ObjC-NSArrays bridge into the same backing storage. However in my testing, they don't have the same backing storage even in ObjC. The only way I see to make this test pass would be to check if the NSDictionary is empty, and native bridge it to the empty singleton. But that would just fail assertions on a different line, since this test requires that the Dictionary is bridged in a non-native way.

I'm very confused by the compilation error the smoke test is giving.

The linux error isn't surprising; I expected to get the `#if runtime(objc)` checks wrong.

Commit:
Some minor cleanups in HashedCollections:

* rename bridgingStorage to bridged
* add a missing fixLifetime
* remove useless temporary

Commit:
update comments in HashedCollections to reflect new design

Gankro commented:
(can reproduce OSX smoke failure; it appears to be a bug in master based on the same crash appearing in everyone's PRs)

Gankro commented:
Rebased onto crashy master.

Added tons of comments.

I'm aware the commit messages are suboptimal; I'll be squashing/cleaning those up when this design is signed off on.

Gankro commented:
Benchmark results: looks like some of the optimized regressions are gone, but there are some new unoptimized regressions. Probably just needs some TLC with inline/final annotations, as I've shifted around implementations a lot.

Bridging performance probably didn't improve because I forgot to implement toll-free bridging the other way. Oops! (fixing now)

Commit:
add new toll-free bridge fast-paths to HashedCollections

Gankro commented:
@swift-ci Please benchmark

Gankro commented:
@swift-ci Please smoke test

Gankro commented:
Correctness-wise, looks like we just need to settle the intent of EqualityTest_Empty.

Gankro commented:
@swift-ci Please benchmark

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -4421,44 +4377,14 @@ internal enum _Variant${Self}Storage<${TypeParametersDecl}> : _HashStorage {
-
+internal struct _Native${Self}Index<${TypeParametersDecl}> : Comparable {
   @_versioned
   internal var offset: Int
```
What is this `offset` thing?  It could use a comment.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -4483,61 +4409,11 @@ extension _Native${Self}Index {
-  internal var allKeys: _HeapBuffer<Int, AnyObject>
-
-  /// Index into `allKeys`
+  /// Index into `allKeys`, a property on the parent collection.
```
What's a "parent collection?"

dabrahams replied:
OK, my point is that it doesn't have a well-defined meaning and should be fixed.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -4602,6 +4468,7 @@ elif Self == 'Dictionary':
 ${SubscriptingWithIndexDoc}
 public struct ${Self}Index<${TypeParametersDecl}> :
   Comparable {
+  // FIXME(swift-3-indexing-model): rewrite the fixme to reflect the new index model
```
I don't understand this comment.

dabrahams commented on validation-test/stdlib/Set.swift:
```diff
@@ -138,7 +138,7 @@ func isNativeSet<T : Hashable>(_ s: Set<T>) -> Bool {
 func isNativeNSSet(_ s: NSSet) -> Bool {
   let className: NSString = NSStringFromClass(type(of: s)) as NSString
-  return className.range(of: "NativeSetStorage").length > 0
+  return className.range(of: "NativeSetBridgingStorage").length > 0
```
I realize this isn't part of the change, but comparing class name strings *really* seems like the wrong way to do this test.  If it isn't wrong, there should be a comment explaining why not.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -454,7 +454,10 @@ public struct Set<Element : Hashable> :
   }
 
-  // TODO: swift-3-indexing-model - add docs
+  /// Returns the index after the given one. 
```
Actually, we don't need to do this anymore unless there's something you need to change about the docs that come from the protocol.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -454,7 +454,10 @@ public struct Set<Element : Hashable> :
+  /// Returns the index after the given one. 
+  ///
+  /// If this is the last valid index, returns endIndex.
+  /// Aborts on endIndex.
```
That should just be `- Precondition: i != endIndex`.  We don't normally document the response to precondition violations; we want the flexibility to change it later.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2063,15 +2069,15 @@ extension Dictionary where Key : Equatable, Value : Equatable {
+        else {
+          return false
         }
-        index = lhsNative.index(after: index)
```
Good job eliminating this useless computation

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2063,15 +2069,15 @@ extension Dictionary where Key : Equatable, Value : Equatable {
         }
-        index = lhsNative.index(after: index)
-        return false
+
```
FYI, in general I am wary of `guard` and these compound conditions; IMO they often don't actually improve readability of the code, I find.  I don't have a strong opinion in this case, but I ask that when you do use `guard` that you convince yourself that the code has improved.

dabrahams commented on test/stdlib/TestUserInfo.swift:
```diff
@@ -20,15 +20,29 @@ import StdlibUnittest
+struct SubStruct: Equatable {
     var i: Int
     var str: String
+
```
This looks like it might be changing what we're testing, since the default for classes (and thus, presumably, id-as-Any boxes) is to compare by object identity.
In any case, the commit message (at least) should describe what practical effect this change has.  Was the test checking the wrong thing?

Gankro replied:
Yeah I determined the test wanted to check for value-equality, and not object-equality. It's literally the only way this assertion makes sense: https://github.com/apple/swift/blob/master/test/stdlib/TestUserInfo.swift#L105

Unless the test *really* intended to rely on the pointer-stability-even-from-different-bridgings fact, in which case it's just testing for something we no longer guarantee.

dabrahams replied:
OK.  Just please try to spell that kind of thing out in commit messages and/or comments

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2454,26 +2449,57 @@ internal struct _${Self}StorageHeader<${TypeParameters}> {
+
+  internal var _body: StorageHeader
+
+  // This API is unsafe and needs a `_fixLifetime` in the caller.
```
We generally try to expose a `withXXX` method to make `_fixLifetime`s less necessary.

Gankro replied:
These are here on master; I don't feel like changing this is in scope for this PR.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2454,26 +2449,57 @@ internal struct _${Self}StorageHeader<${TypeParameters}> {
+  }
+
+  internal static let empty: RawBuffer = {
+    let buffer = Builtin.allocWithTailElems_1(RawBuffer.self,
```
IIRC for `Array` our empty singleton is created in the data segment so there's no runtime work to get it at all.  IMO we should be following that pattern (perhaps in a follow-on PR).

Gankro replied:
That requires modification of the runtime, right? (agreed out of scope, I'll add a FIXME).

dabrahams replied:
I don't remember what it requires, sadly :wink:

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2551,42 +2587,67 @@ final internal class _Native${Self}Storage<${TypeParameters}> :
     let initializedEntries = _UnsafeBitMap(
         storage: _initializedHashtableEntriesBitMapStorage, bitCount: capacity)
-    let keys = _keys
+    let keys = _body.keys.assumingMemoryBound(to: Key.self)
```
Don't we want to sink the `assumingMemoryBound` call into `_body.keys`?

Gankro replied:
_body doesn't know the right type, because it's on RawBuffer. Had to do weird things in here because I wasn't sure what the implication of wrapping self in a NativeStorage during deinit would be.

dabrahams replied:
> _body doesn't know the right type, because it's on RawBuffer. 

Ah, OK.

>Had to do weird things in here because I wasn't sure what the implication of wrapping self in a NativeStorage during deinit would be.

Strange; I'm not sure I understand.

Gankro replied:
If we're in the destructor of a class, can we pass self to other things? What does that even mean?

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -3272,7 +3333,8 @@ final internal class _Native${Self}BridgingStorage<${TypeParametersDecl}>
     get {
       if let ref = _stdlib_atomicLoadARCRef(object: _heapBufferBridgedPtr) {
-        return unsafeDowncast(ref, to: BridgedNativeStorage.Storage.self)
+        return BridgedNativeStorage.Storage(_buffer: unsafeDowncast(ref, 
```
Formatting rule for the standard library: "if you have an open delimiter on a line and its closing delimiter doesn't fit on the same line, break the line after the open delimiter."  That means we break after the first parenthesis above.

I haven't been looking for violations of this rule elsewhere, so you should do an audit.

Gankro replied:
Thanks for the tip!

dabrahams replied:
Still not addressed

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2454,26 +2449,57 @@ internal struct _${Self}StorageHeader<${TypeParameters}> {
-@_versioned
-final internal class _Native${Self}Storage<${TypeParameters}> :
-  ManagedBuffer<_${Self}StorageHeader<${TypeParameters}>, UInt8> {
+internal class _RawNative${Self}Buffer : 
```
Normally, we've been calling the classes "XXXStorage" and the structs that manage those classes "XXXBuffer".  It looks like you're inverting things here.  I think it should be switched back.

Gankro replied:
Ah interesting. This emerged naturally from the old Storage becoming a struct that just wraps these buffers. I wanted the Dictionary to be a thing that has storages (so _NativeStorage and _CocoaStorage are at the same level). However eager bridging will kill the _CocoaStorage so maybe not a big deal?

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2455,8 +2455,15 @@ internal struct _${Self}StorageHeader {
 
-internal class _RawNative${Self}Buffer : 
-  ManagedBuffer<_${Self}StorageHeader, UInt8> {
+/// An instance of this class has all `${Self}` data tail-allocated.
```
Doc comments start with a summary.   Please see API Guidelines.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2455,8 +2455,15 @@ internal struct _${Self}StorageHeader {
-  ManagedBuffer<_${Self}StorageHeader, UInt8> {
+/// An instance of this class has all `${Self}` data tail-allocated.
+/// Enough bytes are allocated to hold the bitmap for marking valid entries,
+/// keys, and values. The data layout starts with the bitmap, followed by the
```
Is all this information about layout an implementation detail, or is it exposed for direct use by other internal components?  If the former, it shouldn't be in a doc comment.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2487,20 +2494,305 @@ internal class _RawNative${Self}Buffer :
+    _sanityCheckFailure("Only initialize these by calling create")
+  }
+
+#if _runtime(_ObjC)
```
Consider breaking all of this out into a separate extension.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2487,20 +2494,305 @@ internal class _RawNative${Self}Buffer :
+  func countByEnumerating(
+    with state: UnsafeMutablePointer<_SwiftNSFastEnumerationState>,
+    objects: UnsafeMutablePointer<AnyObject>?, count: Int
+  ) -> Int {
```
I think I implemented a generic version of this that works for all `Sequence`s conforming to `NSFastEnumeration` or whatever.  You might want to look around to see if you need this.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2487,20 +2494,305 @@ internal class _RawNative${Self}Buffer :
+%end
+}
+
+/// Adds the `Key: Hashable` constraint to _NativeTyped
```
I don't understand the above comment.

Also, it doesn't follow the API guidelines for doc comments.

Gankro replied:
Can you elaborate on the API guidelines? Just that I used a doc comment rather than a "normal" one?

dabrahams replied:
There is lots of detail if you expand https://swift.org/documentation/api-design-guidelines/#write-doc-comment.  Specifically, this one doesn't follow "For all other declarations, describe what the declared entity *is*."

Doc comments should describe APIs; other comments describe implementation details.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2513,6 +2805,47 @@ internal struct _Native${Self}Storage<${TypeParameters}> {
   internal var _buffer: RawBuffer
 
+  /// Creates a Storage with a buffer that is typed, but doesn't understand
+  /// Hashing. Mostly for bridging; prefer `create(capacity:)`.
```
I don't understand what this means.

Gankro replied:
It's `create(capacity:)` that's available when `K: !Hashable`. It must be a different ctor because we instantiate a different class as the buffer.

Gankro replied:
Many operations on the hash table don't actually involve hashing at all. Indexing and iterating being notable examples. You just walk over the structure. In our particular case, we create a `NativeStorage<AnyHashable, AnyHashable>`, and use the indices from `NativeStorage<K, V>` to do lookups in it when doing bridging to objc.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2513,6 +2805,47 @@ internal struct _Native${Self}Storage<${TypeParameters}> {
 
+  /// Creates a Storage with a buffer that is typed, but doesn't understand
+  /// Hashing. Mostly for bridging; prefer `create(capacity:)`.
+  internal static func createUnhashable(capacity: Int) -> Storage {
```
Internal names should be underscored.

dabrahams reviewed (changes requested):
I'm starting to lose my focus about halfway through reviewing 7eaa8752cd6cb308b0012e2936dca6920d1d148c, but I think I've given you plenty to work on for the time being.  Overall it looks like excellent work; thanks for this!

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -1172,15 +1167,13 @@ extension Set {
       }
 
-      if lhsNative.count != rhsNative.count {
+      if lhsNative._count != rhsNative._count {
```
I think this was part of @austinzheng's patch. I'll admit, I don't really fully grok when _'s are considered appropriate or not. Especially for types which are already implementation detail. Seems fairly inconsistent?

Gankro replied:
What does it mean for a field to be an implementation detail, when the containing type is entirely implementation detail? Just that the other parts that use it shouldn't directly make use of that particular implementation detail? Why aren't we just taking advantage of Swift's privacy system for this?

To be clear, all of the Storage/Buffer types, I believe, could be marked fileprivate with no impact.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -175,17 +123,29 @@ import SwiftShims
+//   toll-free bridged to NSDictionary by returning its 
+//   `_RawNativeDictionaryBuffer`.  However most of the actual implementation 
+//   of `NSDictionary` comes from the `_NativeDictionaryBuffer<K, V>`, which is 
+//   the type most `_RawNativeDictionaryBuffer`s have.
```
Yeah I'm maybe trying to stuff too much information in here. I'll clean this up.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -175,17 +123,29 @@ import SwiftShims
+//   Complete bridging is performed when the first access to elements happens. 
+//   This is done by constructing a 
+//   `_NativeDictionaryStorage<AnyObject, AnyObject>` containing a
+//   `_TypedNativeBuffer<AnyObject, AnyObject>`. Since AnyObject isn't
```
"Raw" here is trying to draw analogy to UnsafeRawPointers. They both exist to allow type-agnostic memory manipulation. Does that analogy seem incorrect?

I don't like _TypedNativeBuffer as a name; I just shoe-horned it in. The best example I could find was _ContiguousArrayStorage1, which serves a similar "I needed another thing here but I don't really want it" purpose.

atrick replied:
A "raw" native dictionary would allow heterogeneous value types (without wrapping them in an existential) as long as they are all layout compatible. Since that's not what we're doing I don't think it's appropriate to call it "raw".

Gankro replied:
`TypedBuffer` exists to allow a `NativeStorage<AnyObject, AnyObject>` to be created. RawBuffer can't run destructors because it doesn't have type info, and `NativeBuffer` has to have a `K: Hashable` constraint to fully implement `NSDictionary`.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2495,38 +2482,62 @@ collections = [
+
 /// Header part of the native storage.
-internal struct _HashedContainerStorageHeader {
+internal struct _${Self}StorageHeader<${TypeParameters}> {
```
Yeah I went back on forth on this a few times; always happy to delete more types but definitely better for a followup.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2495,38 +2482,62 @@ collections = [
-}
 
-% for (Self, a_self, TypeParametersDecl, TypeParameters, AnyTypeParameters, Sequence, AnySequenceType) in collections:
+  // Placeholder values until the struct can be initialized.
```
Yes, although I'm unsure what the UB implications for the empty singleton is if it doesn't initialize these at all. It *should* be guaranteed to never access them, if that matters. Just initialize them to point at 0x01?

dabrahams replied:
WFM

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2495,38 +2482,62 @@ collections = [
+  internal var _body: StorageHeader
 
+  /// The number of _logical elements_ the ${a_self} can store. Not to be
+  /// confused with `capacity`, the total number of raw `UInt8` bytes stored
```
This (now deleted) comment was struggling around the inherited capacity from ManagedBuffer; now that we don't inherit from them, this is no longer a thing. (yay!)

dabrahams replied:
yay indeed.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2628,196 +2658,156 @@ final internal class _Native${Self}StorageImpl<${TypeParameters}> {
-  internal typealias Owner = _Native${Self}StorageOwner<${TypeParameters}>
-  internal typealias StorageImpl = _Native${Self}StorageImpl<${TypeParameters}>
+// TODO: conditionally conform to _HashStorage for better compile-time safety.
+// TODO: conditionally conform to CustomStringConvertible.
```
I didn't add these. I'm pretty sure the first one is complaining that we aren't ever asserting NativeStorage: HashStorage, because that only holds when K: Hashable. Not sure why CustomStringConvertible is interesting.

The whole HashStorage thing is kinda dubious and would be removed with eager bridging anyway, though. 

Delete the comments?

dabrahams replied:
That's fine, I guess

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2628,196 +2658,156 @@ final internal class _Native${Self}StorageImpl<${TypeParameters}> {
-  var capacity: Int {
-    return buffer._capacity
+#if _runtime(_ObjC)
+  func bridgingStorage() -> _Native${Self}BridgingStorage<${TypeParameters}> {
```
I renamed/killed the Bridged* types in a later commit. Happy to rename this method to.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2628,196 +2658,156 @@ final internal class _Native${Self}StorageImpl<${TypeParameters}> {
-    }
-    nonmutating set(newValue) {
-      buffer._count = newValue
+  /// A textual representation of `self`.
```
To avoid massive code duplication, I needed to hoist most of the NativeStorage code out of the `K: Hashable` extension so that they're always available. Not all of it needed moving, but I tried to just catch everything that didn't need `K: Hashable`.

What's the motivation of different extensions for different conformances?

dabrahams replied:
It makes it clear what role is being served by each  API.

I understand your need to hoist; it's just generally better, especially in a complex patch like this one, to leave as much text as possible where it is.  Anyway, point made, I presume.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2628,196 +2658,156 @@ final internal class _Native${Self}StorageImpl<${TypeParameters}> {
   @inline(__always)
   internal func key(at i: Int) -> Key {
-    _sanityCheck(i >= 0 && i < capacity)
+    _sanityCheck(i >= 0 && i < _capacity)
```
Yes, I believe the old assertion was in error (and always too large, because it was in terms of bytes). With the new design it's impossible to get this wrong; _capacity is right.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2628,196 +2658,156 @@ final internal class _Native${Self}StorageImpl<${TypeParameters}> {
-    (values + toEntryAt).initialize(to: (from.values + at).move())
-    from.initializedEntries[at] = false
-    initializedEntries[toEntryAt] = true
+
```
Wow great catch.

Gankro replied:
Wow great catch!

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -4153,15 +4119,15 @@ internal enum _Variant${Self}Storage<${TypeParametersDecl}> : _HashStorage {
   /// - parameter offset: The offset of the element that will be deleted.
   /// Precondition: there should be an initialized entry at offset.
-  internal mutating func nativeDeleteImpl(
+  internal mutating func nativeDelete(
```
cc @austinzheng ?

dabrahams replied:
To be clear, I am not opposed to the rename, but I prefer not to see renames as a part of other commits, because it makes it harder to evaluate changes.

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -4240,22 +4203,19 @@ internal enum _Variant${Self}Storage<${TypeParametersDecl}> : _HashStorage {
     // set storage.  Request unique storage.
-    let (reallocated, _) = ensureUniqueNativeStorage(nativeStorage.capacity)
-    if reallocated {
-      nativeStorage = asNative
```
I believe this is no longer relevant with the double-indirection model removed. It was probably fixing up some redundant encoding? That or it was gibberish.

dabrahams replied:
Satisfy yourself that it's OK and I'll be satisfied 😄

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -1671,18 +1664,12 @@ public struct Dictionary<Key : Hashable, Value> :
   public init(minimumCapacity: Int) {
     _variantStorage =
-      .native(_NativeStorage.Owner(minimumCapacity: minimumCapacity))
+      .native(_NativeStorage.create(minimumCapacity: minimumCapacity))
```
That's certainly acceptable for a non-public API.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -136,19 +86,17 @@ import SwiftShims
 // Index Invalidation
 // ------------------
 //
+// FIXME? not sure if this guarantee is worth making, as it restricts
```
I realize now that they don't uniformly do that because we sometimes have parenthesized goop before the colon.  But there's always a colon.  You don't need the `?` IMO.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -175,17 +123,29 @@ import SwiftShims
 //
 // `Dictionary<K, V>` bridges to `NSDictionary` iff both `K` and `V` are
-// bridged.  Otherwise, a runtime error is raised.
+// bridged.  Otherwise, a runtime error is raised. This bridging is always
```
Well, make sure whatever you write is reflected in the code. ;-)

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2983,45 +2964,45 @@ struct _Native${Self}Storage<${TypeParametersDecl}> :
     _ value: Value, forKey key: Key
   ) -> (inserted: Bool, memberAfterInsert: Value) {
     _sanityCheckFailure(
       "don't call mutating methods on _Native${Self}Storage")
```
Well, you're responsible for the PR now, so you should figure it out and/or ask him.

dabrahams commented:
on Thu Oct 20 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> Gankro commented on this pull request.
>
>>          return true
>        }
>
> -      if lhsNative.count != rhsNative.count {
> +      if lhsNative._count != rhsNative._count {
>
> What does it mean for a field to be an implementation detail, when the
> containing type is entirely implementation detail? Just that the other
> parts that use it shouldn't directly make use of that particular
> implementation detail? 

Yes.

> Why aren't we just taking advantage of Swift's privacy system for
> this?

Because we can't use private/fileprivate due to the way the standard
library is compiled.  Ask Jordan if you want more detail.  I don't
understand why, but it's a restriction we have always had to live with.


> To be clear, all of the Storage/Buffer types, I believe, could be
> marked fileprivate with no impact.

Would that it were so!

-- 
-Dave

Commit:
rename Buffer <---> Storage in HashedCollections to match conventions

Commit:
remove _'s from some variables that don't need it

Commit:
rename storage type to be more useful

Commit:
fallout of renaming on tests which rely on internals

Some of these are kinda dubious, but I think this would be better
addressed as part of eager bridging, which will invalidate the concept
most of these are checking for.

Commit:
remove nonsensical identity tests from test suite

The test was verifying that two independently constructed
empty cocoa sets somehow had the same identity. Not sure why this used
to hold, but it's definitely not something we guarantee.

Gankro commented:
@swift-ci Please Smoke Test

Gankro commented:
@swift-ci Please smoke test

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -84,10 +87,55 @@ import SwiftShims
 // ------------------
 //
-// FIXME? not sure if this guarantee is worth making, as it restricts
+// FIXME: decide if this guarantee is worth making, as it restricts
```
This should explicitly cite the alternative.  A link to a robinhood description would help.

Gankro replied:
Sounds good. As far as links go, much of the literature of RHH'ing is bogged down in weird stuff that no real implementation does. Would it be acceptable to link to Rust's implementation notes?

e.g. https://github.com/rust-lang/rust/blob/ac919fcd9d4a958baf99b2f2ed5c3d38a2ebf9d0/src/libstd/collections/hash/map.rs#L70-L178

dabrahams replied:
Something wrong with [this](http://www.sebastiansylvan.com/post/robin-hood-hashing-should-be-your-default-hash-table-implementation/)?  If so, by all means, use the rust thing.  The point here is not to give detail, but merely to be clear about why the restriction is an important limitation.

Gankro replied:
From a quick skim, It makes the common mistake of using tombstones. Other things seem solid.

Commit:
Properly if-def out bridging code.

Commit:
Update comment to clarify why it's interesting, link a relevant impl.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -107,31 +155,42 @@ import SwiftShims
+// `Dictionary<K, V>` bridges to `NSDictionary` in O(1)
+// but may incur an allocation depending on the following conditions:
+//
+// * If the Dictionary is freshly allocated without any elements, then it
```
Is there a difference between “freshly allocated without any elements” and “empty?”  If not, please say ”empty.”  If so, please  be clear about what the distinction is.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2475,30 +2516,41 @@ internal class _RawNative${Self}Buffer:
 
   internal init(_doNotCallMe: ()) {
-    _sanityCheckFailure("Only initialize these by calling create")
+    _sanityCheckFailure("Only create this by using the `empty` singleton")
```
Why is this init needed?  A comment should say.  Also, this should probably call `Builtin.unreachable` if that's not already part of `_sanityCheckFailure`

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2475,30 +2516,41 @@ internal class _RawNative${Self}Buffer:
 
 #if _runtime(_ObjC)
+  //
+  // NSSelf implementation, assuming Self is the empty singleton
```
This makes it look like `NSSelf` is an actual type.  Maybe it should be `NS${Self}` instead?

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2475,30 +2516,41 @@ internal class _RawNative${Self}Buffer:
+  //
+
+  /// Get the NSEnumator implementation for self. NativeSelfBuffer
+  /// overloads this to give NativeSelfNSEnumerator proper type parameters.
```
Ditto, please `${Self}` where appropriate for clarity

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2518,6 +2570,9 @@ internal class _RawNative${Self}Buffer:
     objects: UnsafeMutablePointer<AnyObject>?, count: Int
   ) -> Int {
+    // Even though we never do anything in here, we need to update the
+    // state so that people know we actually ran.
```
“people?”

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2619,7 +2670,7 @@ internal class _TypedNative${Self}Buffer<${TypeParameters}> :
 
   override internal init(_doNotCallMe: ()) {
-    _sanityCheckFailure("Only initialize these by calling create")
+    _sanityCheckFailure("Only create this by calling Storage.createUnhashable")
```
Why is this `init` needed? A comment should say. Also, this should probably call `Builtin.unreachable` if that's not already part of `_sanityCheckFailure`

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2458,9 +2499,9 @@ internal struct _${Self}StorageHeader {
-/// more interesting NSSelf impls.
+/// keys, followed by the values.
+//
+// See the top level docs for more details on this type
```
“top level docs” is ambiguous.  It might send me searching in the project's `docs/` directory.  Ditto every time this term is used.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2648,12 +2700,15 @@ final internal class _Native${Self}Buffer<${TypeParametersDecl}> :
+  // NSSelf bridging:
+
+  /// All actual functionality comes from Storage/Dictionary, which are
+  /// just newtype wrappers around a RawNativeSelfBuffer.
```
`newtype` is evocative of something, but we don't have that feature and omitting the word wouldn't change the meaning. Don't make the reader wonder.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2786,6 +2840,14 @@ final internal class _Native${Self}Buffer<${TypeParametersDecl}> :
 }
 
+
+/// This type provides most of the actual functionality for Dictionary,
```
Sentence fragment please, e.g. "A wrapper around a RawNative${Self}Buffer that provides most of the functionality for ${Self}"

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2823,7 +2886,7 @@ internal struct _Native${Self}Storage<${TypeParameters}> {
 
   /// Given a capacity and uninitialized RawBuffer, completes the 
-  /// initialization and produces a full Storage.  
+  /// initialization and produces a properly formed Storage.  
```
Is there such a thing as an improperly formed Storage?

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2883,6 +2945,7 @@ internal struct _Native${Self}Storage<${TypeParameters}> {
     return _buffer._initializedHashtableEntriesBitMapStorage
   }
 
+  // just a convenience for _keys and _values
```
This comment doesn't tell me what the thing *is*, and its implementation is sufficiently scary that I want to know.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2904,14 +2967,21 @@ internal struct _Native${Self}Storage<${TypeParameters}> {
   }
 %end
 
+  /// Wrap a fully initialized RawBuffer
```
Please expand the [guidelines on doc comments](https://swift.org/documentation/api-design-guidelines/#write-doc-comment) and make these conform

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2856,9 +2919,8 @@ internal struct _Native${Self}Storage<${TypeParameters}> {
-  /// The number of _logical elements_ the ${a_self} can store. Not to be
-  /// confused with `capacity`, the total number of raw `UInt8` bytes stored
-  /// by the `ManagedBuffer`.
+  // Forwarding the individual fields of the buffer in various forms
```
Consider using a swift-divider for these section-division comments.  Look for `//===---` in the source for examples

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -2923,11 +2993,15 @@ internal struct _Native${Self}Storage<${TypeParameters}> {
     return res
   }
 
+  /// Gets the key at the given Index, and then bridges it.
```
You need a paragraph break after the summary.

dabrahams replied:
This should be “Returns the key at the given Index, bridged.“  Again, see the guidelines.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -3435,12 +3511,13 @@ final internal class _SwiftDeferredNS${Self}<${TypeParametersDecl}>
-  // operations on it.
-  //
-  // Do not access this property directly.
+  /// This stored property should be stored at offset zero.  We perform atomic
```
Not suitably arranged to be a doc comment; please fix.

Gankro replied:
There isn't much to say about it on its own; I'll just change it to a "normal" comment.

dabrahams reviewed (changes requested):
Good start; needs some refinement.

dabrahams reviewed (commented):
Looks like I already left plenty of review commentary on this commit which hasn't been addressed.  Is there new stuff I should look at here?

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -3707,7 +3774,7 @@ internal enum _Variant${Self}Storage<${TypeParametersDecl}> : _HashStorage {
 
       let oldNativeStorage = asNative
-      let newNativeStorage = NativeStorage.create(minimumCapacity: minimumCapacity)
+      var newNativeStorage = NativeStorage.create(minimumCapacity: minimumCapacity)
```
What changed that required mutability of this class reference?  A similar question repeats below.

Gankro replied:
Going from a class to a struct made the compiler suddenly start caring about mutability.

dabrahams reviewed (changes requested):
Mostly looks straightforward, but there are some comments I left here that don't appear to have been addressed.

dabrahams reviewed (changes requested):
My one comment on this commit wasn't addressed.

dabrahams reviewed (approved):
The existence of these tests is just wrong and we should (at least) open a radar to remove their dependence on standard library details.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -1349,10 +1349,10 @@ public func _setDownCast<BaseValue, DerivedValue>(_ source: Set<BaseValue>)
     switch source._variantStorage {
     case _VariantSetStorage.native(let storage):
-      let bridgingStorage = storage.bridgingStorage()
+      let bridged = storage.bridged()
```
This constant doesn't seem to help the code; I'd just fold it into the expression below.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -1349,10 +1349,10 @@ public func _setDownCast<BaseValue, DerivedValue>(_ source: Set<BaseValue>)
       return Set(
         _immutableCocoaSet:
-        unsafeBitCast(bridgingStorage, to: _NSSet.self))
+        unsafeBitCast(bridged, to: _NSSet.self))
```
We should be rooting out all `unsafeBitCast` usages in the standard library.  They are almost always the wrong way to do something.

Gankro replied:
Can you elaborate on what the correct code for this might be? I'm still learning all the different cast idioms here. Gut: changed .bridged() to return _NSSet.

dabrahams reviewed (changes requested):
Merely a few nits.

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -5124,16 +5124,27 @@ extension ${Self} {
     }
   }
 
+  /// Gets the native Dictionary hidden inside this NSDictionary
```
Guidelines say:

Describe what a function or method _does_ and what it _returns_, omitting null effects and Void returns:

```swift
/// Inserts `newHead` at the beginning of `self`.
mutating func prepend(_ newHead: Int)

/// Returns a `List` containing `head` followed by the elements
/// of `self`.
func prepending(_ head: Element) -> List

/// Removes and returns the first element of `self` if non-empty;
/// returns `nil` otherwise.
mutating func popFirst() -> Element?
```

Gankro replied:
Just to be clear: s/Gets/Returns is the only issue here?

dabrahams commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -5124,16 +5124,27 @@ extension ${Self} {
-      // re-wrap the native storage.
-      return ${Self}<${TypeParameters}>(_nativeStorage: bridgingStorage.nativeStorage)
+
+    // Try all three NSSelf impls that we currently provide.
```
Would it be better to cast to a common base class and if it succeeds, unconditionally call a method that does the rest of the work?  Just askin'.

Gankro replied:
I thought about it but there's no compression that can be done here (I think):

* _NSDeferred isn't connected to the Storage types beyond NSString itself

* After checking that you *could* cast to RawStorage, but the body of HashableTypedStorage still needs to do yet-another-type-check (for the type params), so you're not really saving a branch, while adding a dynamic call.

dabrahams reviewed (changes requested):
Guidelines again

dabrahams reviewed (approved):
OK, I looked at the diff locally and it appears to be entirely mechanical.

dabrahams reviewed (approved):
This is fine; it would be nice to see a little rationale as to why these properties don't need the leading underscore in the commit message

dabrahams reviewed (changes requested):
If you're going to use an enigmatic word like "sneaky" in the commit message, please explain it down below.  I think you mean "tests that touch stdlib implementation details"?

dabrahams reviewed (approved):
Nice.  Phew, this PR was a lot of work on your part!

dabrahams commented:
on Thu Nov 03 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> From a quick skim, It makes the common mistake of using
> tombstones. Other things seem solid.

That should not be a problem, as they are incidental to the point,
right?

Gankro commented on stdlib/public/core/HashedCollections.swift.gyb:
```diff
@@ -3152,7 +3152,8 @@ internal struct _Native${Self}Buffer<${TypeParameters}> {
   where ${'Key' if Self == 'Dictionary' else 'Element'} : Hashable
 {
-  internal typealias HashTypedStorage = _Native${Self}Storage<${TypeParameters}>
+  internal typealias HashTypedStorage = 
```
It could be, but honestly it shows up in like one or two places so it might as well be a bit verbose for clarity. It's mostly here to put all the GYB gunk in one place.

dabrahams commented:
on Thu Nov 03 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> A small one: emptying out a dictionary doesn't regress it to the empty storage. Only constructing it
> without any capacity (e.g. init()) does.

OK, I think I was being to nitpicky; the language is fine in context.

-- 
-Dave

dabrahams commented:
on Thu Nov 03 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> Can you elaborate on what the correct code for this might be? I'm still learning all the different
> cast idioms here. Gut: changed .bridged() to return _NSSet.

I can't find this comment/commit from the GitHub web interface (!) so
it's hard to say, but if you can have `.bridged()` return `_NSSet`
without introducing any `unsafeBitCast`s, that's almost certainly the
right answer!

-- 
-Dave

Commit:
more comment cleanups

Commit:
Fix incorrect bitcasts, refactor code to guard against this error.

Commit:
leave FIXME for future dictionary bucaneers

Commit:
cleanup initializers in HashedCollections

Commit:
actually used the cached keys/values pointer

Gankro commented:
@swift-ci Please smoke test

Gankro commented:
@swift-ci please benchmark

Gankro commented:
@dabrahams I believe I've addressed all of your comments

(also fixed a mem-safety bug, yay!)

Gankro commented:
@swift-ci Please smoke test

dabrahams reviewed (commented):
What was happening _before_ you did this?!

dabrahams commented:
on Thu Nov 03 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> @dabrahams I believe I've addressed all of your comments

W00t! I trust that it's ready to go as soon as the tests pass; feel free
to merge then.

Commit:
remove buffer header to eliminate optional traps

Gankro commented:
@swift-ci please smoke test

Gankro commented:
@swift-ci Please benchmark

Gankro commented:
@swift-ci Please smoke test

Gankra committed:
remove useless constant redefinition

Gankro commented:
@swift-ci Please smoke test

Gankro commented:
@swift-ci Please smoke test

Gankro commented:
@swift-ci Please smoke test

Gankro commented:
Heck yeah, those are some perf numbers I can get behind! 😎

Gankro commented:
Oh god all tests passed I'm scared what do I do can I just push that big green button is this real oh god oh god oh god 😱

Gankro commented:
@bob-wilson here's my perf analysis:

The two biggest regressions are the result of removing the bridging cache that Dictionaries implicitly had. This means that:

```
for _ in 0..<100 {
  myDict as NSDictionary
}
```

creates 100 independent `_SwiftDeferredNSDictionary` with independent "bridged buffer" caches, where previously it would just return the same one over and over. If this is considered problematic, we can potentially add a weak pointer caching a `_SwiftDeferredNSDictionary` to `_RawNativeDictionaryStorage` (or maybe only  `_HashableTypedNativeDictionaryStorage`).

I haven't dug into the other regressions, but it's worth noting that this PR changed the representation of a lazy bridged dictionary (`_VariantDictionaryStorage.cocoa`) to be a class, where previously it was a struct to allow `lazy var` to be used. Reverting this might be the right solution; I didn't obsess over this since eager bridging may just delete all that code.

dabrahams commented:
Congratulations @gankro and @austinzheng !

bob-wilson commented:
@Gankro Since you and Dave already discussed the performance tradeoff and concluded that this is the right thing to do, I'm not going to second guess that. If you don't end up deleting the code with eager bridging, can you come back and take another look at the performance later?

dabrahams commented:
@Gankro @bob-wilson Actually I am unclear on why this should cause any regressions at all, so maybe we should discuss that.

Gankro commented:
This should cause a regression (at least in synthetic benchmarks) because:

* It eliminates the repeated-bridging cache (possibly only a problem for micro benchmarks that bridge in a loop?). We can in principle add this back, at the cost of storing a weak or strong pointer (depending on how eagerly you want to free that cached allocation) in every `_RawNative${Self}Storage`.

* It requires a `_SwiftDeferredNSDictionary` to be eagerly allocated to wrap the the two `_Native${Self}Buffers`. We can in principle eliminate this by making that hypothetical pointer into a strong pointer, and having the NSDictionary implementation branch on whether K/V is verbatim bridgeable *for every operation*.

The former is potentially desirable, the latter is probably a huge waste. 

This is an unavoidable consequence of making Dictionary into a single-allocation data structure. A lot of optimizations were built around there being two allocations. In effect, we've taken the cost of allocating an extra class from every dictionary, and pushed it into every bridge. I consider that an acceptable trade-off.

Gankro commented:
Mental reworking: we can move the verbatim bridgeable check to Storage construction time, and just fork `_HashableTypedNative${Self}Storage` into two types. The one that exists now, and one that has the slot for a strong pointer to the bridged buffer. This means `Dictionary<Int, Int>` would store a potentially unused pointer, while `Dictionary<MyClass, MyClass>` wouldn't. Much better.

Gankro commented:
Followup from private conversation:

@dabrahams we definitely need/want to add a new Storage class so that every method doesn't have to be guarded by "am I verbatim bridgeable", unless I'm missing something.

Also I'm remembering now that we dicussed this design like a month ago; I believe we concluded that storing a strong pointer in the storage is unacceptable, because that is an easy source of memory leaks. However whatever we return to ObjC must store a strong pointer. Note that the solution implemented by this PR is the solution used by Array; see _SwiftDeferredNSArray (which I modelled this solution after).

dabrahams commented:
on Thu Nov 10 2016, Alexis Beingessner <notifications-AT-github.com> wrote:

> Followup from private conversation:
>
> @dabrahams we definitely need/want to add a new Storage class so that
> every method doesn't have to be guarded by "am I verbatim bridgeable",
> unless I'm missing something.

That check gets compiled down to nothing any place where the static type
of the element is visible, so that cost is dramatically less than you
might assume.

> Also I'm remembering now that we dicussed this design like a month
> ago; I believe we concluded that storing a strong pointer in the
> storage is unacceptable, because that is an easy source of memory
> leaks. However whatever we return to ObjC must store a strong
> pointer. Note that the solution implemented by this PR is the solution
> used by Array; see _SwiftDeferredNSArray (which I modelled this
> solution after).

Eh, I think I'll need an interactive refresher on that discussion.

-- 
-Dave
